crates/contract/src/tee/proposal.rs

```diff
@@ -50,70 +50,60 @@ impl CodeHashesVotes {
-    pub image_hash: MpcDockerImageHash,
-    pub docker_compose_hash: LauncherDockerComposeHash,
-    pub added: BlockHeight,
+    pub(crate) image_hash: MpcDockerImageHash,
+    pub(crate) docker_compose_hash: LauncherDockerComposeHash,
+    pub(crate) added: Timestamp,
 }
 /// Collection of whitelisted Docker code hashes that are the only ones MPC nodes are allowed to
 /// run.
-#[near(serializers=[borsh, json])]
-#[derive(Clone, Default, Debug, PartialEq, Eq)]
-pub struct AllowedDockerImageHashes {
+#[near(serializers=[json])]
+#[derive(Clone, Default, Debug, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub(crate) struct AllowedDockerImageHashes {
     /// Whitelisted code hashes, sorted by when they were added (oldest first). Expired entries are
     /// lazily cleaned up during insertions and TEE validation.
     allowed_tee_proposals: Vec<AllowedMpcDockerImage>,
 }
 
 impl AllowedDockerImageHashes {
     /// Checks if a Docker image hash is still valid (not expired).
-    fn is_image_hash_valid(
-        entry: &AllowedMpcDockerImage,
-        tee_upgrade_deadline_duration_blocks: u64,
-        current_block_height: BlockHeight,
-    ) -> bool {
-        entry.added + tee_upgrade_deadline_duration_blocks >= current_block_height
-    }
+    fn valid_entries(&self, tee_upgrade_deadline_duration: Duration) -> Vec<AllowedMpcDockerImage> {
+        let current_time = Timestamp::now();
 
-    /// Removes all expired code hashes and returns the number of removed entries.
-    /// Ensures that at least one (the latest) proposal always remains in the whitelist.
-    pub fn cleanup_expired_hashes(
-        &mut self,
-        current_block_height: BlockHeight,
-        tee_upgrade_deadline_duration_blocks: u64,
-    ) -> usize {
-        // Find the first non-expired entry, but never remove the last one
-        let expired_count = self
-            .allowed_tee_proposals
+        self.allowed_tee_proposals
             .iter()
-            .position(|entry| {
-                Self::is_image_hash_valid(
-                    entry,
-                    tee_upgrade_deadline_duration_blocks,
-                    current_block_height,
-                )
+            .enumerate()
+            .filter(|(i, _entry)| {
+                // The latest entry is always valid.
+                let Some(successor) = self.allowed_tee_proposals.get(i + 1) else {
+                    return true;
+                };
+                let grace_period_deadline = successor.added + tee_upgrade_deadline_duration;
+
+                current_time <= grace_period_deadline
             })
-            .unwrap_or(self.allowed_tee_proposals.len());
-
-        // Never remove all proposals; always keep at least one (the latest)
-        let expired_count = expired_count.min(self.allowed_tee_proposals.len().saturating_sub(1));
-
-        self.allowed_tee_proposals.drain(0..expired_count);
+            .map(|(_i, entry)| entry)
+            .cloned()
+            .collect()
```

Comment:
This looks incorrect. Why do we return the current entry if the successor is valid?

If the logic is "skip the latest entry, and check the grace period of the rest" it should instead be something like:

```suggestion
            .rev()
            .skip(1)
            .filter(|entry| {
                let grace_period_deadline = entry.added + tee_upgrade_deadline_duration;

                current_time <= grace_period_deadline
            })
            .cloned()
            .collect()
```

Note that I'm returning in reverse order here. I don't think order matters, and it seems more sensible to return the most recent one first anyway.

Let me know if I'm missing something here. Otherwise it would also be good to add test coverage for this.

Reply:
This is precisely the bug that is being removed. Here you are defining the expiration deadline as the `tee_upgrade_deadline_duration` plus the entry time of the potentially stale hash itself, not the entry time of its successor which is what should be starting the ticking timer to invalidate it.

Reply:
> This looks incorrect. Why do we return the current entry if the successor is valid?

Isn't that the whole point of having a grace period?